worker/changestream/eventmultiplexer/eventmultiplexer.go

```diff
@@ -0,0 +1,302 @@
+	stream   Stream
+	logger   Logger
+	clock    clock.Clock
+
+	subscriptions      map[uint64]*subscription
+	subscriptionsByNS  map[string][]*eventFilter
+	subscriptionsAll   map[uint64]struct{}
+	subscriptionsCount uint64
+
+	// (un)subscription related channels to serialize adding and removing
+	// subscriptions. This allows the queue to be lock less.
+	subscriptionCh   chan subscriptionOpts
+	unsubscriptionCh chan uint64
+}
+
+// New creates a new EventMultiplexer that will use the Stream for events.
+func New(stream Stream, clock clock.Clock, logger Logger) (*EventMultiplexer, error) {
+	queue := &EventMultiplexer{
+		stream:             stream,
+		logger:             logger,
+		clock:              clock,
+		subscriptions:      make(map[uint64]*subscription),
+		subscriptionsByNS:  make(map[string][]*eventFilter),
+		subscriptionsAll:   make(map[uint64]struct{}),
+		subscriptionsCount: 0,
+
+		subscriptionCh:   make(chan subscriptionOpts),
+		unsubscriptionCh: make(chan uint64),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &queue.catacomb,
+		Work: queue.loop,
+	}); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return queue, nil
+}
+
+// Subscribe creates a new subscription to the event queue. Options can be
+// provided to allow filter during the dispatching phase.
+func (e *EventMultiplexer) Subscribe(opts ...changestream.SubscriptionOption) (changestream.Subscription, error) {
+	// Get a new subscription count without using any mutexes.
+	subID := atomic.AddUint64(&e.subscriptionsCount, 1)
+
+	sub := newSubscription(subID, func() { e.unsubscribe(subID) })
+	if err := e.catacomb.Add(sub); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	select {
+	case <-e.catacomb.Dying():
+		return nil, e.catacomb.ErrDying()
+	case e.subscriptionCh <- subscriptionOpts{
+		subscription: sub,
+		opts:         opts,
+	}:
+	}
+
+	return sub, nil
+}
+
+// Kill stops the event queue.
+func (e *EventMultiplexer) Kill() {
+	e.catacomb.Kill(nil)
+}
+
+// Wait waits for the event queue to stop.
+func (e *EventMultiplexer) Wait() error {
+	return e.catacomb.Wait()
+}
+
+func (e *EventMultiplexer) unsubscribe(subscriptionID uint64) {
+	select {
+	case <-e.catacomb.Dying():
+		return
+	case e.unsubscriptionCh <- subscriptionID:
+	}
+}
+
+func (e *EventMultiplexer) loop() error {
+	defer func() {
+		for _, sub := range e.subscriptions {
+			sub.close()
+		}
+		e.subscriptions = nil
+		e.subscriptionsByNS = nil
+
+		close(e.subscriptionCh)
+		close(e.unsubscriptionCh)
+	}()
+
+	for {
+		select {
+		case <-e.catacomb.Dying():
+			return e.catacomb.ErrDying()
+
+		case term, ok := <-e.stream.Terms():
+			// If the stream is closed, we expect that a new worker will come
+			// again using the change stream worker infrastructure. In this case
+			// just ignore and close out.
+			if !ok {
+				e.logger.Infof("change stream term channel is closed")
+				return nil
+			}
+
+			changeSet := make(map[*subscription]ChangeSet)
```

Comment:
Again, premature optimisation, but you could drop the need for a map, by storing the changeset on the subscription itself.